polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -0,0 +1,285 @@
+
+  private def transformUpdate(update: NotebookUpdate, toVersion: GlobalVersion): NotebookUpdate = {
+    updateBuffer.getRange(update.globalVersion, toVersion).foldLeft(update) {
+      case (accum, (ver, next)) => accum.rebase(next)
+    }
+  }
+
+  private val subscribers = new ConcurrentHashMap[Int, Subscriber]()
+  private val nextSubscriberId = new AtomicInteger(0)
+
+  def open(name: String, oq: Enqueue[IO, Message]): IO[NotebookRef[IO]] = for {
+    subscriberId <- IO(nextSubscriberId.getAndIncrement())
+    subscriber    = new Subscriber(subscriberId, name, oq)
+    _            <- IO { subscribers.put(subscriberId, subscriber); () }
+  } yield subscriber
+
+  def versions: Stream[IO, (GlobalVersion, Notebook)] = ref.discrete
+
+  class Subscriber(subscriberId: Int, name: String, oq: Enqueue[IO, Message]) extends NotebookRef[IO] {
+    private val lastLocalVersion = new AtomicInteger(0)
+    private val lastGlobalVersion = new AtomicInteger(0)
+
+    private val closeSignal = ReadySignal()
+
+    def setKnownVersions(global: Int, local: Int): IO[Unit] = IO {
+      lastGlobalVersion.set(global)
+      lastLocalVersion.set(local)
+    }
+
+    def lastKnownGlobalVersion: Int = lastGlobalVersion.get()
+
+    // listen for output messages / status updates, and publish them to my client
+    outputMessages.subscribe(1024).interruptWhen(closeSignal())
+      .to(oq.enqueue).compile.drain.unsafeRunAsyncAndForget()
+
+
+    // listen for foreign updates, transform them appropriately, and sink to foreignEdits queue
+    private val updateFiber = updatesTopic.subscribe(1024).interruptWhen(closeSignal()).unNone.filter(_._2 != subscriberId).evalMap {
+      case (globalVersion, _, update) =>
+        val knownGlobalVersion = lastGlobalVersion.get()
+
+        if (globalVersion < knownGlobalVersion) {
+          // this edit should come before other edits I've already seen - transform it up to knownGlobalVersion
+          IO.pure(transformUpdate(update, knownGlobalVersion).withVersions(knownGlobalVersion, lastLocalVersion.get()))
+        } else if (globalVersion > knownGlobalVersion) {
+          // this edit should come after the last global version I've seen - client will transform locally if necessary
+          IO.pure(update.withVersions(globalVersion, lastLocalVersion.get()))
+        } else {
+          // something has gone terribly wrong.
+          IO.raiseError(new IllegalStateException(s"Duplicate global version $globalVersion"))
+        }
+    }.covaryOutput[Message].to(oq.enqueue).compile.drain.start.unsafeRunSync()
+
+    val path: String = IOSharedNotebook.this.path
+
+    override def get: IO[Notebook] = ref.get.map(_._2)
+
+    override def update(update: NotebookUpdate): IO[Int] = {
+      for {
+        versionPromise <- submitUpdate(subscriberId, update)
+        version        <- versionPromise.get
+      } yield version
+    }
+
+    override def close(): IO[Unit] = closeSignal.complete.flatMap(_ => updateFiber.join)
```

Comment:
Just so I understand what this does:

We complete `closeSignal` and wait for `updateFiber` to finish, which will happen soon because we triggered an interruption by completing the signal in the first place. So basically we want to make sure that `updateFiber` has finished before saying we're actually closed (because I guess `updateFiber` might still be doing something).

In that case, should we also be waiting on `outputMessages.subscribe ...` as well?

Reply:
Hmm, it's probably not necessary to wait on `updateFiber`. It just depends on when you consider the Subscriber to be "closed" – the `IO[Unit]` returned by `close()` shouldn't complete before that.